sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala

```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
+                  compressedBlockArray.size)
+                taskMemoryManager.freeUnchecked(block)
+                compressedBlock
+              case None => block
+            }
+          }
+
+          def hasNext: Boolean = bufferedRowIterator.hasNext
+        }
+      }.setName(compressionType + "_" + df.toString).persist(StorageLevel.MEMORY_ONLY)
+
+      val baseRelation: BaseRelation = new BaseRelation with TableScan {
+        override val sqlContext = _sqlContext
+        override val schema = df.schema
+        override val needConversion = false
+
+        override def buildScan(): RDD[Row] = {
+          val numFields = this.schema.length
+          val _compressionType: String = compressionType
+          val _blockSize = blockSize
+
+          cachedRDD.flatMap { rawBlock =>
+            // Optionally decompress block
+            val compressionCodec: Option[CompressionCodec] = if (_compressionType.isEmpty) {
+              None
+            } else {
+              Some(CompressionCodec.createCodec(SparkEnv.get.conf, _compressionType))
+            }
+            val block = compressionCodec match {
+              case Some(codec) =>
+                // Copy compressed block (excluding padding) to on-heap byte array
+                val padding = Platform.getInt(rawBlock.getBaseObject, rawBlock.getBaseOffset)
+                val compressedBlockArray = new Array[Byte](_blockSize)
+                Platform.copyMemory(
+                  rawBlock.getBaseObject,
+                  rawBlock.getBaseOffset + 4,
+                  compressedBlockArray,
+                  Platform.BYTE_ARRAY_OFFSET,
+                  rawBlock.size() - padding)
+
+                // Decompress into MemoryBlock backed by on-heap byte array
+                val compressedBais = new ByteArrayInputStream(compressedBlockArray)
+                val uncompressedBlockArray = new Array[Byte](_blockSize)
+                val cdis = new DataInputStream(codec.compressedInputStream(compressedBais))
+                // compression codec has its own block size and cis.read only reads one compression
+                // codec block, so we need to keep reading until cis is exhausted
+                // TODO: Use DataInputStream.readFully()
+                cdis.readFully(uncompressedBlockArray)
+//                var cumNumRead = 0
+//                var numRead = cis.read(uncompressedBlockArray)
+//                while (numRead > 0) {
+//                  cumNumRead += numRead
+//                  numRead = cis.read(
+//                    uncompressedBlockArray,
+//                    cumNumRead,
+//                    uncompressedBlockArray.length - cumNumRead)
+//                }
+//                cis.close()
+                cdis.close()
+                MemoryBlock.fromByteArray(uncompressedBlockArray)
+              case None => rawBlock
+            }
+
+            val rows = new ArrayBuffer[InternalRow]()
```

Comment:
Couldn't you return an iterator of rows which streams through the block to deserialize the rows lazily? I wonder whether buffering all of these rows is going to be expensive + involve lots of unnecessary object creation.